File: examples/05-production/python/custom_activity_metric.py
from temporalio import activity

# Custom Activity metric. The SDK gives every Activity a MetricMeter wired to the
# same Prometheus runtime (see prometheus_metrics.py), so a custom counter shows
# up alongside the built-in SDK metrics. This is the Python analogue of holding a
# Micrometer Counter in InvoiceActivitiesImpl.


@activity.defn
async def generate_invoice(order_id: str) -> str:
    counter = activity.metric_meter().create_counter(
        "training_invoices_generated_total",
        "Number of invoices generated",
    )
    counter.add(1)
    return f"s3://invoices/{order_id}.pdf"

File: examples/05-production/go/custom_activity_metric.go
package production

import (
	"context"

	"go.temporal.io/sdk/activity"
)

// Custom Activity metric. activity.GetMetricsHandler(ctx) returns the same
// handler the SDK uses, wired to whatever client.MetricsHandler was configured
// (see prometheus_metrics.go), so a custom counter is exported alongside the
// built-in SDK metrics. The Go analogue of holding a Micrometer Counter in
// InvoiceActivitiesImpl.
func GenerateInvoice(ctx context.Context, orderID string) (string, error) {
	activity.GetMetricsHandler(ctx).
		Counter("training_invoices_generated_total").
		Inc(1)
	return "s3://invoices/" + orderID + ".pdf", nil
}
